File: 1418-Fair Distribution of Cookies.java
class Solution {
    int ret = Integer.MAX_VALUE;
    public int distributeCookies(int[] cookies, int k) {
        helper(k, 0, cookies, new int[8]);
        return ret;
        
    }
    public void helper(int k, int cur, int[] arr, int[] plan){
        if(cur == arr.length){
            int mx = 0;
            for(int i : plan)
                mx = Math.max(mx, i);
            ret = Math.min(ret, mx);
            return;
        }
        for(int i = 0; i < k; i++){
            plan[i] += arr[cur];
            helper(k, cur + 1, arr, plan);
            plan[i] -= arr[cur];
        }
    }
}

File: 1418-Fair Distribution of Cookies.py
class Solution:
    def distributeCookies(self, cookies: List[int], k: int) -> int:
        cur = [0] * k
        N = len(cookies)
        ret = float('inf')
        
        def helper(i):
            nonlocal ret
            if i >= N:
                mx = max(cur)
                ret = min(ret, mx)
                return
            if ret <= max(cur): 
                return
            for j in range(k):
                cur[j] += cookies[i]
                helper(i+1)
                cur[j] -= cookies[i]
            
        helper(0)
        return ret
    
    
    def distributeCookies(self, cookies: List[int], k: int) -> int:
        N = len(cookies)
        def helper(i, val, cur):
            if i >= N:
                return True
            
            for j in range(k):
                if cur[j] + cookies[i] > val:
                    continue
                cur[j] += cookies[i]
                if helper(i+1, val, cur):
                    return True
                cur[j] -= cookies[i]
            
            return False
        
        l = 0
        r = sum(cookies)
        while l <= r:
            mid = l + (r - l) // 2
            cur = [0] * k
            if helper(0, mid, cur):
                r = mid - 1
            else:
                l = mid + 1
            
        return l